src/Microsoft.AspNet.Mvc.Core/ActionConstraints/ConsumesAttribute.cs

```diff
@@ -0,0 +1,135 @@
+                if (requestContentType != null &&
+                    !ContentTypes.Any(contentType => contentType.IsSubsetOf(requestContentType)))
+                {
+                    context.Result = new UnsupportedMediaTypeResult();
+                }
+            }
+        }
+
+        public void OnResourceExecuted([NotNull] ResourceExecutedContext context)
+        {
+        }
+
+        private bool IsApplicable(ActionConstraintContext context)
+        {
+            return context.CurrentCandidate.Action.FilterDescriptors.Last(filter => filter.Filter is ConsumesAttribute).Filter == this;
+        }
+
+        public bool Accept(ActionConstraintContext context)
+        {
+            // First check to see if this is applicble
+            if (!IsApplicable(context))
+            {
+                // since this constraint is not applicable it should return true.
+                return true;
+            }
+
+            MediaTypeHeaderValue requestContentType = null;
+            MediaTypeHeaderValue.TryParse(context.RouteContext.HttpContext.Request.ContentType, out requestContentType);
+
+            // If the request content type is null we need to act like pass through.
+            // In case there is a single candidate with a constraint it should be selected.
+            // If there are multiple actions with consumes action constraints this should result in ambiguous exception
+            // unless there is another action without a consumes constraint.
+            if (requestContentType == null)
+            {
+                var isActionWithoutConsumeConstraintPresent = context.Candidates.Any(
+                    candidate => candidate.Constraints == null ||
+                    !candidate.Constraints.Any(constraint => constraint is IConsumesActionConstraint));
+
+                return !isActionWithoutConsumeConstraintPresent;
+            }
+
+            if (ContentTypes.Any(c => c.IsSubsetOf(requestContentType)))
+            {
+                return true;
+            }
+
+            var firstCandidate = context.Candidates[0];
+            if (firstCandidate != context.CurrentCandidate)
+            {
+                // If the current candidate has reached here, its not a match.
+                return false;
+            }
+
+            foreach (var candidate in context.Candidates)
+            {
+                if (candidate == firstCandidate)
+                {
+                    continue;
+                }
+
+                var tempContext = new ActionConstraintContext()
+                {
+                    Candidates = context.Candidates,
+                    RouteContext = context.RouteContext,
+                    CurrentCandidate = candidate
+                };
+
+                if (candidate.Constraints == null || candidate.Constraints.Count() == 0 ||
+                    candidate.Constraints.Any(constraint => constraint is IConsumesActionConstraint &&
```

Comment:
well that is a slippery slope right? if there is a custom implementation of ```IConsumesActionConstraint``` we would not be able to capture that. Not fixing this let me know if you think otherwise.